common/src/main/java/io/netty/util/concurrent/DefaultPromise.java

```diff
@@ -90,36 +92,93 @@ protected DefaultPromise() {
-            return true;
-        }
-        return false;
+        return trySuccess0(result);
+    }
+
+    private boolean trySuccess0(V result) {
+        return trySetValue(result == null ? SUCCESS : result);
     }
 
     @Override
     public Promise<V> setFailure(Throwable cause) {
-        if (setFailure0(cause)) {
-            notifyListeners();
+        if (tryFailure0(cause)) {
             return this;
         }
         throw new IllegalStateException("complete already: " + this, cause);
     }
 
     @Override
     public boolean tryFailure(Throwable cause) {
-        if (setFailure0(cause)) {
-            notifyListeners();
-            return true;
+        return tryFailure0(cause);
+    }
+
+    private boolean tryFailure0(Throwable cause) {
+        return trySetValue(new CauseHolder(checkNotNull(cause, "cause")));
+    }
+
+    private boolean trySetValue(Object value) {
+        EventExecutor executor = executor();
+        boolean inEventLoop = executor.inEventLoop();
+        if (inEventLoop) {
+            final InternalThreadLocalMap threadLocals = InternalThreadLocalMap.get();
+            final int stackDepth = threadLocals.futureListenerStackDepth();
+            if (stackDepth < MAX_LISTENER_STACK_DEPTH) {
+                if (RESULT_UPDATER.compareAndSet(this, null, value) ||
+                        RESULT_UPDATER.compareAndSet(this, UNCANCELLABLE, value)) {
+                    threadLocals.setFutureListenerStackDepth(stackDepth + 1);
+
+                    // Notify the waiters (and so unblock these) before we notify the listeners.
+                    checkNotifyWaiters();
+
+                    try {
+                        notifyListenersNow();
+                    } finally {
+                        threadLocals.setFutureListenerStackDepth(stackDepth);
+                    }
+                    return true;
+                }
+                return false;
+            }
+            // fall-through
+        }
+        return trySetValueLater(executor, value);
+    }
+
+    private boolean trySetValueLater(final EventExecutor executor, final Object value) {
+        // As we will need to offload the notification of the listeners to the EventExecutor, we need to ensure
+        // we can actually at least submit the task the EventExecutor before we will let the caller know its safe to
+        // assume the ownership of the value was "transferred" to the DefaultPromise. This is especially important
+        // if the value will be of type ReferenceCounted as the user may expect to be able to release it in the listener
+        // if the try / set operation succeeded.
+        if (RESULT_UPDATER.compareAndSet(this, null, PRESET) ||
+                RESULT_UPDATER.compareAndSet(this, UNCANCELLABLE, PRESET)) {
+            if (safeExecute(executor, new Runnable() {
+                @Override
+                public void run() {
+                    if (RESULT_UPDATER.get(DefaultPromise.this) == PRESET) {
```

Comment:
@rkapsi yes it could ... that said I think using a `get` and then a `set` if needed is cheaper then a compareAndSet and should be good enough for what we try here.